### PR TITLE
type config with `None` default as `str | None` instead of `Any`

### PR DESCRIPTION
**Pull request by chdsbd**
A config with a default of `None` now returns `str | None` instead of `Any`.

```python
from starlette.config import Config
config = Config()
PRIVATE_KEY = config("PRIVATE_KEY", default=None)

# before
reveal_type(PRIVATE_KEY) # Revealed type is "Any"

# after
reveal_type(PRIVATE_KEY) # Revealed type is "Union[builtins.str, None]"
```

Initially raised as discussion https://github.com/encode/starlette/discussions/1730

### CODE REVIEW

File: tests/test_config.py
Comment (chdsbd) [inline tests/test_config.py:12]: I added some tests for the type annotations. Open to feedback on how to make these nicer

File: tests/test_config.py
Comment (chdsbd) [inline tests/test_config.py]: See https://github.com/encode/starlette/discussions/1731

In a future PR, I'd like to require that the default be either `None` or the same type as the `cast`.

File: tests/test_config.py
Comment (Kludex) [reply tests/test_config.py:12]: Maybe parametrize those:
```python
STR = [config("STR")]
is_str(STR)
```
And move the function definitions outside this test function (above) 🤔

**Review by Kludex — COMMENTED** [review]
I'm not sure if we should take an approach in which we do assertions instead of depending on the type checker.

I've asked around how people do this, not sure if helpful, but there are some ideas: https://twitter.com/marcelotryle/status/1544374655481532422

File: setup.py
Comment (Kludex) [inline setup.py]: Please undo this 😗 
```suggestion
        "typing_extensions>=3.10.0; python_version < '3.10'",
```

File: tests/test_config.py
Comment (Kludex) [inline tests/test_config.py:12]: Maybe adding a docstring explaining a bit that we are relying on mypy? 🙏

**Review by Kludex — COMMENTED** [review]
I added some comments. I'm overall fine with this PR. 🙏

File: requirements.txt
Comment (chdsbd) [inline requirements.txt:12]: I added typing_extensions here because we need a newer version that contains assert_type

Comment (Kludex) [reply requirements.txt:12]: Thanks 🙏

**Review by Kludex — APPROVED** [review]
Thanks @chdsbd 🙌

I'll let merge this in a few days if there are no further comments from other members. 😁

File: tests/test_config.py
Comment (Kludex) [inline tests/test_config.py]:
```suggestion
```

**Pull request merged**
